Add documentation for enumerations (#758) (#924)
graphql-dotnet/graphql-dotnet: docs2/site/docs/getting-started/schema-types.md

}
```

**.NET**

Consider the equivalent `enum` in .NET:

```csharp
public enum Episodes
{
    NEWHOPE  = 4,
    EMPIRE  = 5,
    JEDI  = 6
}
```

Compare the two implementations. GraphQL does not specify backing values for members of its enums. The name of each member _is_ the value.

**GraphQL .NET**

GraphQL.NET provides two methods of defining GraphQL enums.

You can manually create the `EnumerationGraphType`. Advantages of this method:

- The GraphQL enum need not map to a specific .NET `enum`. You could, for instance, build the enum from one of the alternate methods of defining discrete sets of values in .NET, such as classes of constants or static properties

> - Allows descriptions of enum members

Reviewer: Obsolete. `EnumerationGraphType<TEnum>` sets descriptions.

Reviewer 2: I will add this detail but the existing method of defining enum value descriptions is not obsolete in the sense that it's still part of the public API and is not itself marked obsolete. Not everyone will be keen to use the attribute method of defining enum descriptions.

Reviewer: I am sorry, I incorrectly expressed. I just wanted to say that the advantages of this method can be perceived as opposed to the generic version, which did not have such capabilities before (descriptions, deprecations). I agree that it's OK to say about the explicit possibilities of the first method. Then it is worth clarifying that the second method also supports similar capabilities, but through attributes.

> - Allows deprecation of enum members

Reviewer: regarding deprecation: #1102

```csharp
[Description("The Star Wars movies.")]
[Obsolete("Optional. Sets the GraphQL DeprecationReason for the whole enum.")]
public enum Episodes
{
    [Description("Episode 1: The Phantom Menace")]
    [Obsolete("Optional. Sets the GraphQL DeprecationReason for this member.")]
    PHANTOMMENACE = 1,

    [Description("Episode 4: A New Hope")]
    NEWHOPE  = 4,

    [Description("Episode 5: The Empire Strikes Back")]
    EMPIRE  = 5,

    [Description("Episode 6: Return of the Jedi")]
    JEDI  = 6
}

public class EpisodeEnum : EnumerationGraphType<Episodes>
{
}
```

> You can also manually create the `EnumerationGraphType`.  Advantages of this method:

Reviewer: ```suggestion
You can also manually create the `EnumerationGraphType`. Advantages of this method:
```

> - The GraphQL enum need not map to a specific .NET `enum`.  You could, for instance, build the enum from one of the alternate methods of defining discrete sets of values in .NET, such as classes of constants or static properties.

Reviewer: ```suggestion
- The GraphQL enum need not map to a specific .NET `enum`. You could, for instance, build the enum from one of the alternate methods of defining discrete sets of values in .NET, such as classes of constants or static properties.
```

> - You can manually add description and deprecation reasons.  This may be useful if you do not control the source code for the enum.

Reviewer: ```suggestion
- You can manually add descriptions and deprecation reasons. This may be useful if you do not control the source code for the enum.
```

> - Backing values may be any primitive type or string.

Reviewer: ```suggestion
- Backing enum values may be of any type, primitive or not.
```

@andrewtyped This is important.

}
```

**.NET**

Consider the equivalent `enum` in .NET:

```csharp
public enum Episodes
{

>     NEWHOPE  = 4,

Reviewer: ```suggestion
    NEWHOPE = 4,
```

>     EMPIRE  = 5,

Reviewer: ```suggestion
    EMPIRE = 5,
```

>     JEDI  = 6

Reviewer: ```suggestion
    JEDI = 6
```

}
```

> Compare the two implementations. GraphQL does not specify backing values for members of its enums. The name of each member _is_ the value.

Reviewer: Good point, thanks.

**GraphQL .NET**

GraphQL.NET provides two methods of defining GraphQL enums.

> You can use `EnumerationGraphType<TEnum>` to automatically generate values by providing a .NET `enum` for `TEnum`. The `Name` will default to the .NET Type name, which you can override in the constructor.  The `Description` will default to any `System.ComponentModel.DescriptionAttribute` applied to the enum type.  The `DeprecationReason` will default to any `System.ObsoleteAttribute` applied to the enum type.  By default, the name of each enum member will be converted to upper case.  Override `ChangeEnumCase` to change this behavior.  Apply a `DescriptionAttribute` to an enum member to set the GraphQL `Description`.  Apply an `ObsoleteAttribute` to an enum member to set the GraphQL deprecation reason.

Reviewer: ```suggestion
You can use `EnumerationGraphType<TEnum>` to automatically generate values by providing a .NET `enum` for `TEnum`. The `Name` will default to the .NET Type name, which you can override in the constructor. The `Description` will default to any `System.ComponentModel.DescriptionAttribute` applied to the enum type. The `DeprecationReason` will default to any `System.ObsoleteAttribute` applied to the enum type. By default, the name of each enum member will be converted to CONSTANT_CASE. Override `ChangeEnumCase` to change this behavior. Apply a `DescriptionAttribute` to an enum member to set the GraphQL `Description`. Apply an `ObsoleteAttribute` to an enum member to set the GraphQL deprecation reason.
```

> You can use `EnumerationGraphType<TEnum>` to automatically generate values by providing a .NET `enum` for `TEnum`.  The `Name` will default to the .NET Type name, which you can override in the constructor.  The `Description` will default to any `System.ComponentModel.DescriptionAttribute` applied to the enum type.  The `DeprecationReason` will default to any `System.ObsoleteAttribute` applied to the enum type.  By default, the name of each enum member will be converted to upper case.  Override `ChangeEnumCase` to change this behavior.  Apply a `DescriptionAttribute` to an enum member to set the GraphQL `Description`.  Apply an `ObsoleteAttribute` to an enum member to set the GraphQL deprecation reason.

Reviewer: ```suggestion
You can use `EnumerationGraphType<TEnum>` to automatically generate values by providing a .NET `enum` for `TEnum`. The `Name` will default to the .NET Type name, which you can override in the constructor.  The `Description` will default to any `System.ComponentModel.DescriptionAttribute` applied to the enum type.  The `DeprecationReason` will default to any `System.ObsoleteAttribute` applied to the enum type.  By default, the name of each enum member will be converted to upper case.  Override `ChangeEnumCase` to change this behavior.  Apply a `DescriptionAttribute` to an enum member to set the GraphQL `Description`.  Apply an `ObsoleteAttribute` to an enum member to set the GraphQL deprecation reason.
```

```csharp

> 

Reviewer: ```suggestion
```

}
```

**.NET**

> 

Reviewer: ```suggestion
```

}
```

> You can also use the generic version by passing it a .NET `enum` which will populate the values for you (excluding description).  The `Name` will default to the .NET Type name, which you can override in the constructor. By default, the name of each enum member will be converted to upper case. This behavior can be changed by overriding `ChangeEnumCase`.

Reviewer: > excluding description

Obsolete. EnumerationGraphType<TEnum> automatically sets descriptions from attributes.

}
```

> Note that although GraphQL has no use for backing values for enum members, GraphQL.NET uses them anyway. This allows for a more natural mapping to .NET `enum`s or other collections of constants, and avoids coupling your business logic to GraphQL semantics. The backing values are strictly for use on the back end - the client will never see them.

Reviewer: Good point also!

**Resolving Enumerations**

> Fields typed as enumerations are resolved by returning either the name or backing value of one or more of the enum members. In the below examples, notice the identical implementations of the `appearsIn` field for both human graph types. In both implementations, the client receives the enum member names in response to queries on the `appearsIn` field.

Reviewer: > of one or more 

What do you mean?

Author: Clarified this in #1205775f. I meant to show that resolving enums works equally well with list types.

Reviewer: OK

If the field resolves a value which cannot be mapped to one of the enum's legal values, GraphQL.NET will return `null` to the client in the data for the field.

```csharp
public class HumanString
{
    //i.e. "NEWHOPE", "EMPIRE", "JEDI"
    public string[] AppearsIn { get; set; }
}

public class HumanStringType: ObjectGraphType<HumanString>
{
    public HumanStringType()
    {
        Name = "HumanString";
        Field<ListGraphType<EpisodeEnum>>("appearsIn", "Which movie they appear in.");
    }
}

public class HumanInt
{
    //i.e. 4, 5, 6
    public int AppearsIn { get; set; }
}

public class HumanIntType: ObjectGraphType<HumanInt>
{
    public HumanIntType()
    {
        Name = "HumanInt";
        Field<ListGraphType<EpisodeEnum>>("appearsIn", "Which movie they appear in.");
    }
}
```

**Enumeration Arguments**

Enumerations can be used as arguments in GraphQL queries. Consider a query which gets the humans appearing in a specific film:

```graphql

> query HumansAppearingIn($episode: Episode){

Reviewer: ```suggestion
query HumansAppearingIn($episode: Episode!){
```

Author: Resolved in 3f6a4d7

    humans(appearsIn: $episode){
        id
        name
        appearsIn
    }
}

# example query variables:
# {
#   "episode":"NEWHOPE"
# }
```

When GraphQL.NET receives an enum member name as a query argument, the queried field's `ResolveFieldContext` stores the backing value associated with the enum member name in its arguments list. The GraphQL.NET query type which handles the example query may be implemented as:

```csharp
    public class StarWarsQuery : ObjectGraphType<object>
    {
        public StarWarsQuery()
        {
            Name = "Query";

            Field<ListGraphType<HumanType>>(
                "humans",
                arguments: new QueryArguments(

>                     new QueryArgument<EpisodeEnum> 

Reviewer: ```suggestion
                    new QueryArgument<NonNullGraphType<EpisodeEnum>>
```

Author: Resolved in b7d15c8

                        { 
                            Name = "appearsIn", 
                            Description = "An episode the human appears in." 
                        }
                ),
                resolve: context => 
                {

>                     //episode = 4

Reviewer: ```suggestion
                    // episode = 4
```

>                     var episode = context.GetArgument<int>("appearsIn");

Reviewer: It will be useful to show that `var episode = context.GetArgument<Episodes>("appearsIn");` works too.

>                     //full implementation would access data store to get humans by episode.

Reviewer: ```suggestion
                    // full implementation would access data store to get humans by episode.
```
